### ISSUE TITLE
Aggregate Targets (and schemes)

### ISSUE THREAD

**Issue by lordcodes** [tuist/tuist#3388]
## Context 🕵️‍♀️

In our project we used an Aggregate Target (and scheme) so that we could run quality checks within Xcode, SwiftLint and SwiftFormat. The advantage of doing this was that you could easily select it from schemes and that any errors would be surfaced within the Issues Navigator where they can be clicked to quickly jump to the issues and fix them in code. Without this, you would need to run from command line and then work out what was wrong and find the locations manually within the project.

## What 🌱

Unless I have completely failed to find it in the docs and via exploring the API, it doesn't seem to be possible to define Aggregate Targets using Tuist.

## Proposal 🎉

A property on the `Project` to provide a list of Aggregate Targets. They are relatively simple, only really requiring a name and a list of actions (scripts) to add as build phases. Each Aggregate Target needs an auto-generated scheme that allows you to run the target using the 'Run Action'.

**Comment by fortmarek** [comment]
Hi @lordcodes!

Yes, this seems like a sensible addition and there is no real reason not to support it.

Would you like to implement this? I can give you guidance if necessary.

**Comment by lordcodes** [comment]
Sure. I was planning to take a look into the code sometime soon. Will ask
if I have any questions!

**Comment by adellibovi** [comment]
Out of curiosity @lordcodes, is there any specific reason why you would prefer an aggregate target instead of build phase for things like SwiftLint?

**Comment by lordcodes** [comment]
Running SwiftLint and SwiftFormat takes time, I would rather not run them
on every single build. I would prefer to run them only when we want to. The
best way I found of doing this was an aggregate target.

As an alternative I tried making a scheme that ran them as a pre-action but
it didn’t seem to surface issues in the Issue Navigator. I may have been
wrong on that though.

**Comment by luispadron** [comment]
I agree with you @lordcodes, I've found it difficult to run a tool like `swiftlint` or `swiftformat` which operate on a set of sources outside of the target, thus repeating work for every target in the project.

Adding the action to a scheme, does **not** surface errors either, this only seems to work for target actions which is very odd on Xcodes part.

It seems like the correct way to use things like `swiftlint` & `swiftformat` is passing the path to only the input sources in the target action, this way it only lints and formats the files for the given target. However, not all tools support this and I think if we could make this easier somehow it would be a great addition.

**Comment by lordcodes** [comment]
Glad to hear it will be useful to someone else.

**Comment by kubajakowski** [comment]
> Would you like to implement this? I can give you guidance if necessary.

@fortmarek If you don't have plans for implementing it in near future I can take a look but I'll need a guidance on what exactly has to be done to implement new product type support. Can you help?

**Comment by mmatoszko** [comment]
@fortmarek Also interested in this. How can I help? :smile:

**Comment by fortmarek** [comment]
Hey @kubajakowski @mmatoszko!

Sorry for the late response 😞 

Fortunately, `XcodeProj` which is used for generating already [supports aggregate targets](https://github.com/tuist/XcodeProj/blob/main/Sources/XcodeProj/Objects/Targets/PBXAggregateTarget.swift). 

The method for generating targets is [here](https://github.com/tuist/tuist/blob/ef07576961aee66fa46b9dc393df0c4c41e1e970/Sources/TuistGenerator/Generator/TargetGenerator.swift#L53) and used [here](https://github.com/tuist/tuist/blob/ef07576961aee66fa46b9dc393df0c4c41e1e970/Sources/TuistGenerator/Generator/ProjectDescriptorGenerator.swift#L108). 

The issue is that that method return `PBXNativeTarget` which `PBXAggregateTarget` does not inherit from. In the `ProjectDescriptorGenerator`, you will need to create a new method for generating aggregate targets specifically instead.

You will either need to update the API of `Target` in `ProjectDescription` (see [here](https://github.com/tuist/tuist/blob/ef07576961aee66fa46b9dc393df0c4c41e1e970/Sources/ProjectDescription/Target.swift)) and in `TuistGraph` (see [here](https://github.com/tuist/tuist/blob/ef07576961aee66fa46b9dc393df0c4c41e1e970/Sources/TuistGraph/Models/Target.swift)) and add it as a [new Product](https://github.com/tuist/tuist/blob/ef07576961aee66fa46b9dc393df0c4c41e1e970/Sources/TuistGraph/Models/Target.swift#L20). Or most likely since aggregate target are more constrained in what you can do with them, you will need to create a new model completely and add a new field in `Project` called `aggregateTargets: [AggregateTarget]`. 

I am more inclined to the second option but it's worth doing a research on both.

I am happy to discuss this more in Slack as well 🤞

**Comment by lordcodes** [comment]
I feel like the better way to run tools like SwiftFormat and SwiftLint will become SPM plugins that can be ran from command line or via the Xcode 14 UI.

**Comment by danyf90** [comment]
> I feel like the better way to run tools like SwiftFormat and SwiftLint will become SPM plugins that can be ran from command line or via the Xcode 14 UI. For this reason I'm not sure adding support for Aggregate Targets is as necessary.

Makes sense, let's close this for now then. We can reopen it in case we find another use case for them 👍

**Comment by yamschikov-vs** [comment]
We want to include aggregate target for analyse unused code https://github.com/peripheryapp/periphery. But can't do this =(

**Comment by oozoofrog** [comment]
> I feel like the better way to run tools like SwiftFormat and SwiftLint will become SPM plugins that can be ran from command line or via the Xcode 14 UI. For this reason I'm not sure adding support for Aggregate Targets is as necessary.

When using SwiftLint or peripheral, it would be nice to be able to use the results to make them more visible as warnings or errors in the xcode ide, rather than just listing and checking the results. It would be nice to have aggregate applied for that.

**Comment by fortmarek** [comment]
We would appreciate any contributions to this, we can guide you how to implement this 🙏

**Comment by apascual** [comment]
@fortmarek I can give it a try, where should I get started? :)

**Comment by fortmarek** [comment]
I actually provided some pointers [here](https://github.com/tuist/tuist/issues/3388#issuecomment-1129645454), let me know if you need more 🤞

**Comment by devdrey** [comment]
Hello,
Another use case for aggregate targets we use a lot is execution of custom scripts that generate code for us.

We currently migration to tuist and lacking this capability a lot. We found workarounds to use other target types which do not contain any sources. But this is rather a hack and has its limitations.

We have a quite old tech stack and can not jet switch to Swift packages for our +100 repos. hence SwiftPackage Pligins re no option yet without major effort in the migration process.

So +1000 to agregate target support!

**Comment by apascual** [comment]
> 

Hi, I had now a look to this and tried your second approach. Unfortunately I noticed that in `XcodeProj` sources, `PBXAggregateTarget` inherits a lot of unusable properties from `PBXTarget`, like for example `productType` or `packageProductDependencies`. My understanding is that aggregate targets can only have `name`, `buildPhases` and `buildRules`, how should we deal with this?

**Comment by MontakOleg** [comment]
@apascual @fortmarek @devdrey hi guys!

I have tried to implement initial support for aggregate targets in two ways: 
1. using separate `AggregateTarget` model
2. using existing Target with `product = .aggregateTarget`.

The second approach was much easier. Will be happy to discuss details in PR: https://github.com/tuist/tuist/pull/5266

**Comment by fortmarek** [comment]
Reopening this as this is being worked on, thanks @MontakOleg!

**Comment by fedulvtubudul** [comment]
There seems to be a new use case for aggregate targets. "Build Tool plug-ins" were introduced with Xcode 15, and these plugins always run before other build phases in same target.

One of the uses I can see for myself is to use OpenAPI generator build tool plugin for generating API client from a swagger spec, and a separate target for downloading an up-to-date spec file before running the generation step.

**Comment by paulso1410** [comment]
Hello, there's a plan to to finish this development? as I see, the https://github.com/tuist/tuist/pull/5266 is closed. We want to use Periphery with Xcode Support using the aggregate target. I can try to help, but I don't know if there's a stopper

**Comment by MontakOleg** [comment]
Hi @paulso1410, we use regular framework target to run SwiftLint/Periphery, it works fine:

```
extension Target {
    public static func makeAggregateTarget(name: String, scripts: [TargetScript] = []) -> Target {
        Target(
            name: name,
            platform: .macOS,
            product: .framework,
            bundleId: "",
            sources: [],
            scripts: scripts
        )
    }
}

...

// and then use it in Project.swift like this
private func aggregateTargets() -> [Target] {
    [
        .makeAggregateTarget(
            name: "Swiftlint",
            scripts: [
                .pre(
                    path: "./Scripts/run_swiftlint_via_mint.sh",
                    name: "Swiftlint",
                    basedOnDependencyAnalysis: false
                )
            ]
        ),
        .makeAggregateTarget(
            name: "Periphery",
            scripts: [
                .pre(
                    path: "./Scripts/run_periphery_via_mint.sh",
                    name: "Periphery",
                    basedOnDependencyAnalysis: false
                )
            ]
        )
    ]
}
```

**Comment by shahzadmajeed** [comment]
@MontakOleg I'm curious why these scripts are not in a "build phase" and what is the benefit of using a MacOS target instead?

**Comment by paulso1410** [comment]
> Hi @paulso1410, we use regular framework target to run SwiftLint/Periphery, it works fine:
> 
> ```
> extension Target {
>     public static func makeAggregateTarget(name: String, scripts: [TargetScript] = []) -> Target {
>         Target(
>             name: name,
>             platform: .macOS,
>             product: .framework,
>             bundleId: "",
>             sources: [],
>             scripts: scripts
>         )
>     }
> }
> 
> ...
> 
> // and then use it in Project.swift like this
> private func aggregateTargets() -> [Target] {
>     [
>         .makeAggregateTarget(
>             name: "Swiftlint",
>             scripts: [
>                 .pre(
>                     path: "./Scripts/run_swiftlint_via_mint.sh",
>                     name: "Swiftlint",
>                     basedOnDependencyAnalysis: false
>                 )
>             ]
>         ),
>         .makeAggregateTarget(
>             name: "Periphery",
>             scripts: [
>                 .pre(
>                     path: "./Scripts/run_periphery_via_mint.sh",
>                     name: "Periphery",
>                     basedOnDependencyAnalysis: false
>                 )
>             ]
>         )
>     ]
> }
> ```

Awesome! thanks, I'm going to try it

**Comment by MontakOleg** [comment]
@shahzadmajeed, we run such tools in pre-commit hook + give ability to run them manually from Xcode. The goal – to not spend a few seconds on every build.
Inspired by https://www.merowing.info/improve-build-times-by-extracting-3rd-party-tooling-to-processing-script/

About `.macOS` vs `.iOS` target – I don't think that there any practical difference.

**Comment by AlexanderGermek** [comment]
@MontakOleg did u have problems with unit tests? when im add periphery:
```
    .target(
      name: "\(TargetName.periphery)",
      destinations: .macOS,
      product: .framework,
      bundleId: "",
      sources: [],
      scripts: [
        .pre(
          script: "./Tools/periphery/run_periphery.sh",
          name: "Run Periphery",
          basedOnDependencyAnalysis: false
        )
      ],
      settings: .settings(base:
                            ["ASSETCATALOG_COMPILER_APPICON_NAME": "",
                             "DEFINES_MODULE": "NO"])
    )
```

and when run my scheme (cmd + B) i have building errors:

No such module "NameOfMainTarget" in @testable import "NameOfMainTarget" in my unit tests.

my run_periphery.sh part of script:
```
xcrun --sdk macosx "$PERIPHERY_BIN" scan --verbose \
  --project "$ROOT_DIR/RBO.xcworkspace" \
  --schemes RBO \
  --retain-public \
  --retain-unused-protocol-func-params \
  --clean-build \
  --format xcode \
  -- \
  build \
  -sdk iphonesimulator \
  -destination "platform=iOS Simulator,id=$SIMULATOR_UDID" \
  CODE_SIGNING_ALLOWED=NO \
  ENABLE_BITCODE="NO" \
  COMPILER_INDEX_STORE_ENABLE="YES" \
  INDEX_ENABLE_DATA_STORE="YES" \
  CODE_SIGN_IDENTITY="" \
  CODE_SIGNING_REQUIRED=NO \
  EXPANDED_CODE_SIGN_IDENTITY="" \
  EXPANDED_CODE_SIGN_IDENTITY_NAME="" \
  EXPANDED_PROVISIONING_PROFILE=""
```

**Comment by MontakOleg** [comment]
Hi @AlexanderGermek

Yes, we had similar issues after updating to Periphery 3.

`run_periphery_xcode.sh`:
```bash
# Some variables set by Xcode breaks build.
# We want to run periphery in fresh env without variables set by Xcode.
# It can be achieved by using
#   env -i "$SRCROOT"/Scripts/run_periphery_xcode.sh
# OR
#   by unsetting explicitly the variables that breaks the build
unset SWIFT_INSTALL_MODULE_FOR_DEPLOYMENT
unset TEST_FRAMEWORK_SEARCH_PATHS
unset TEST_LIBRARY_SEARCH_PATHS

periphery scan
```

**Comment by AlexanderGermek** [comment]
@MontakOleg  THK U very much! It works! I've been looking for this for a very long time 💪
